criu/util.c

```diff
@@ -1804,3 +1804,75 @@ int run_command(char *buf, size_t buf_size, int (*child_fn)(void *), void *args)
 
 	return fret;
 }
+
+/*
+ * This function cuts sub_path from the path.
+ * 1) It asumes all relative paths given are relative to "/":
+ * 	/a/b/c is the same as a/b/c
+ * 2) It can handle paths with multiple consequent slashes:
+ * 	///a///b///c is the same as /a/b/c
+ * 3) It always returns relative path, with no leading slash:
+ * 	get_relative_path("/a/b/c", "/") would be "a/b/c"
+ * 	get_relative_path("/a/b/c", "/a/b") would be "c"
+ * 	get_relative_path("/", "/") would be ""
+ * 4) It can handle paths with single dots:
+ * 	get_relative_path("./a/b", "a/") would be "b"
+ * 5) Note ".." in paths are not supported and handled as normal directory name
+ */
+char *get_relative_path(char *path, char *sub_path)
```

Comment:
Function get_relative_path has two main features:
 
1) It works with almost any complex paths (".." is exception, ".." are handled as simple directory names because resolving them as actual "parent directory" link may require fs accesses);
2) It does NO allocation and NO memory writes, returned relative path pointer is a shifted path pointer.

In contrast path_simplify need to allocate memory and need to move chars. In criu file/mount code we construct a lot of paths, to be robust we would need to allocate and write again when simplifying each constructed path.

So we already have many path construction code and patching all of it is a big deal I believe. Integrating path_simplify into get_relative_path looks like a bad solution because the latter is(/would be) called in many places and I like that it does not add memory complexity to the overall algorithm.

Reply:
Yeah, sounds reasonable. I do generally prefer to always simplify and clean paths before processing them further in the code so I'm biased.